Build Class for Report Charts

aloftus23 opened the pull request:
## 🗣 Description ##

Adds chart code using matplotlib in Class structure. fixes #83 

## 💭 Motivation and context ##

Required to build matplotlib charts in the report.

## 🧪 Testing ##

Passes pytests and pre-commits

## ✅ Checklist ##

<!-- Remove any of the following that do not apply. -->
<!-- Draft PRs should have one or more unchecked boxes. -->
<!-- If you're unsure about any of these, don't hesitate to ask. -->
<!-- We're here to help! -->

- [x] This PR has an informative and human-readable title.
- [x] Changes are limited to a single goal - _eschew scope creep!_
- [x] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [x] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [x] All new and existing tests pass.

aloftus23 committed:
Created chart file and class

Includes these functions:
- pie
- stacked_bar
- h_bar
- line_chart

aloftus23 committed:
add matplotlib to install_requires

aloftus23 committed:
fix color being defined twice and line breaks

aloftus23 committed:
add __init__ function to Chart class

dav3r commented:
`src/.DS_Store` should not be committed to the repository - please remove it from this PR.

dav3r commented on src/pe_reports/charts.py:
```diff
@@ -0,0 +1,162 @@
+        height = self.height
+        name = self.name
+        plt.rcParams.update({"figure.max_open_warning": 0})
+        catName = df.columns[0]
+        Val_1Name = df.columns[1]
+        df = df.sort_values(by=Val_1Name, ascending=False)
+        Category_column = df[catName]
+        Value_column = df[df.columns[1]]
+        labels = Category_column
```
You have a mix of naming styles with your variables (e.g. `catName` vs. `Category_column`).  You should choose one style and use it consistently.  We don't have it codified in our [style guide](https://github.com/cisagov/development-guide/blob/develop/style-guide.md), but we try to use the [Pythonic `lower_case_with_underscores` naming convention](https://www.python.org/dev/peps/pep-0008/#function-and-variable-names) for functions and variables.

aloftus23 replied:
@dav3r Addressed in this [commit](https://github.com/cisagov/pe-reports/pull/106/commits/44b345ab38b499d181dbc95e3dfe43e001650cbb).

dav3r replied:
That looks much better! 👍

dav3r commented on src/pe_reports/charts.py:
```diff
@@ -0,0 +1,162 @@
+
+        # Only show the label when the value is greater than 10%.
+        def autopct(pct):
+            """Get autopct."""
```
This docstring should be improved - it's not particularly useful.

aloftus23 replied:
@dav3r Addressed in this [commit](https://github.com/cisagov/pe-reports/pull/106/commits/874c4c1d018162acd7be4458e78a7d90062657ea).

dav3r commented on src/pe_reports/charts.py:
```diff
@@ -0,0 +1,162 @@
+            frameon=False,
+        )
+        plt.subplots_adjust(left=0.2, wspace=0.2)
+        plt.gcf().set_size_inches(width / 2.54, height / 2.54)
```
The `2.54` magic number is repeated throughout this class.  You should consider replacing it with a constant.

aloftus23 replied:
@dav3r Addressed in this [commit](https://github.com/cisagov/pe-reports/pull/106/commits/16db49cbf1c57112f2024e3264651e4d9c26ba32).

dav3r replied:
Excellent, thanks 👍

dav3r commented on src/pe_reports/charts.py:
```diff
@@ -0,0 +1,162 @@
+        fig, ax = plt.subplots()
+        ax.spines.right.set_visible(False)
+        ax.spines.top.set_visible(False)
+        plt.plot(df[df.columns[0]], Value_column, label="Dark Web Mentions")
```
This chart has the label ("Dark Web Mentions") hard-coded.  I'd recommend passing the label in as a variable in order to generalize this function.

aloftus23 replied:
@dav3r Addressed in the [commit](https://github.com/cisagov/pe-reports/pull/106/commits/5b2a900ef2bb1ddbaf8bc2e9dd7c485e7be8930d).

dav3r commented on src/pe_reports/charts.py:
```diff
@@ -0,0 +1,162 @@
+        plt.gcf().set_size_inches(width / 2.54, height / 2.54)
+        plt.xticks(fontsize=7)
+        plt.yticks(fontsize=7)
+        plt.gca().set_ylabel("Mentions count", labelpad=10, fontdict={"size": 8})
```
Same issue as above with a hard-coded label.

aloftus23 replied:
@dav3r Addressed in the [commit](https://github.com/cisagov/pe-reports/pull/106/commits/5b2a900ef2bb1ddbaf8bc2e9dd7c485e7be8930d).

dav3r reviewed (changes requested):
Please take a look at my suggestions and let me know what you think.

aloftus23 committed:
Remove DS_Store and add it to .gitignore

aloftus23 committed:
Fix variable naming conventions

Change all variable names to snake case.

dav3r commented on .gitignore:
```diff
@@ -9,3 +9,4 @@ __pycache__
 .pytest_cache
 .python-version
 *.egg-info
+.DS_Store
```
`.DS_Store` should be moved to a separate section called `## macOS ##` and since we alphabetize whenever possible, that section should be placed above the `## Python ##` section.

aloftus23 replied:
@dav3r Addressed in this [commit](https://github.com/cisagov/pe-reports/pull/106/commits/dcccb110fd6a07e1b8302988e69d7d2c4cb9c3fd).

dav3r commented:
Reminder that you should update this PR checklist above and check all relevant boxes for items that you have done.  If there are any items that you plan to do, but haven't done them yet, leave those unchecked until you do them.  You can delete the checklist items that are not applicable to this PR (e.g. if you haven't added any new TODOs or if the documentation doesn't need to be updated).

Before the PR is merged, all remaining checkboxes should be checked.

aloftus23 committed:
Improve autopct docstring

Change the docstring to "Get percentages for the pie chart slices."

aloftus23 committed:
Replace repeated conversion factor with a constant

Replace the cm to inches conversion factor of 2.54 with the constant CM_CONVERSION_FACTOR.

aloftus23 committed:
Generalize line_chart function by removing hard-coded labels

Replace hard-coded values with the x_label and y_label variables.

aloftus23 committed:
Create macOS section in .gitignore and alphebitize

Edit .gitignore to follow the suggested standard.

dav3r reviewed (approved):
Thanks for taking the time to make the changes I suggested! 💎

aloftus23 commented:
> Thanks for taking the time to make the changes I suggested! 💎

Thanks for walking me through it!

cduhn17 commented on src/pe_reports/charts.py:
```diff
@@ -0,0 +1,175 @@
+
+        # Only show the label when the value is greater than 10%.
+        def autopct(pct):
+            """Get percentages for the pie chart slices."""
```
```suggestion
            """Get percentages for the pie chart slices and only 
            show the label when the value is greater than 10%.
            """
```

cduhn17 commented on src/pe_reports/charts.py:
```diff
@@ -0,0 +1,175 @@
+        labels = category_column
+        plt.gca().axis("equal")
+
+        # Only show the label when the value is greater than 10%.
```
```suggestion
        
```

cduhn17 reviewed (changes requested):
@aloftus23 , 
Minor changes see suggestions.

aloftus23 committed:
Combine comment and docstring

No need to have the comment and the docstring so combined them.

Co-authored-by: cduhn17 <87447403+cduhn17@users.noreply.github.com>

aloftus23 committed:
Remove unnecessary comment

Remove comment that was moved into the docstring

Co-authored-by: cduhn17 <87447403+cduhn17@users.noreply.github.com>

aloftus23 committed:
Fix lint errors and make docstring one line

aloftus23 commented:
@cduhn17 Adjusted the docstring. Let me know what you think.

cduhn17 reviewed (approved):
@aloftus23 , 

Looks good!

cduhn17 committed:
Merge branch 'develop' into AL-charts-WIP
